### PR TITLE
Add support for async hooks. Closes gh-43.

### PR DESCRIPTION
**Pull request by paulmillr**
I wasn’t sure about adding `async.js` dep, so I implemented simple async foreach fn.

### CODE REVIEW

**Comment by paulmillr** [comment]
@visionmedia bump. please look at this. i’m making cool component plugin, will be awesome if this was released before NY.

File: lib/builder.js [hunk 1]
```diff
@@ -533,34 +533,36 @@ Builder.prototype.buildType = function(type, fn, process){
...
-          if (null == val) {
-            debug('read file %s', path);
-            fs.readFile(path, 'utf8', function(err, str){
-              if (err) return fn(err);
-              done(null, process(self, file, str));
-            });
-            return
-          }
+    self.hook('before ' + type, self, function(error) {
+      if (error) throw error;
```
Comment (tj) [inline lib/builder.js]: oh noes we need to delegate these

File: test/hooks.js [hunk 2]
```diff
@@ -16,9 +16,11 @@ var Builder = require('..')
   , read = fs.readFileSync;
 
 function ejsPlugin(builder) {
-  builder.hook('before scripts', function(pkg){
+  console.log(123)
```
Comment (tj) [inline test/hooks.js]: meow

File: test/hooks.js [hunk 3]
```diff
@@ -16,9 +16,11 @@ var Builder = require('..')
...
+
+  builder.defineHook('before scripts', function(pkg, callback){
```
Comment (tj) [inline test/hooks.js]: I'd like to keep this as .hook() since .defineHook is just the private implementation

Comment (tj) [inline test/hooks.js]: It would be nice to do the arity check as well so `callback` isn't necessary for sync stuff

File: lib/builder.js [hunk 4]
```diff
@@ -736,13 +738,24 @@ Builder.prototype.hook = function(name, fn){
...
+  var completed = 0;
+
   for (var i = 0, len = fns.length; i < len; ++i) {
-    arg = fns[i](arg);
+    fn = fns[i];
+    fn(arg, function(error) {
+      completed++;
+      if (error) lastError = error;
+      if (completed != len) return;
+      callback(lastError);
```
Comment (tj) [inline lib/builder.js]: is this the async for each art you were mentioning? I'd definitely rather not add big deps but we could easily add a little setting to Batch to make it optionally sequential or a separate `each` package that is more declarative, that's probably taken in npm though

Comment (paulmillr) [inline lib/builder.js]: yep

File: lib/builder.js (hunk 1 above)
Comment (paulmillr) [inline lib/builder.js]: how exactly?

Comment (tj) [inline lib/builder.js]: i cant see in this diff but we must have a callback up there somewhere

**Comment by paulmillr** [comment]
1. Arity check is now done, sync fns can be used.
2. Removed `console.log`
3. Delegation is done instead of `throw error`.

**Comment by tj** [comment]
squashed/merged!

**Comment by paulmillr** [comment]
thanks dude

**Pull request closed without merging**